src/engine/TransitivePath.cpp

```diff
@@ -658,119 +449,158 @@ void TransitivePath::computeTransitivePathRightBound(
+                                       size_t startSideCol,
+                                       size_t targetSideCol) {
+  size_t rowIndex = 0;
+  for (auto const& [node, linkedNodes] : hull) {
+    for (Id linkedNode : *linkedNodes) {
+      table.emplace_back();
+      table(rowIndex, startSideCol) = node;
+      table(rowIndex, targetSideCol) = linkedNode;
+
+      rowIndex++;
+    }
+  }
 }
 
 // _____________________________________________________________________________
-std::shared_ptr<TransitivePath> TransitivePath::bindRightSide(
-    std::shared_ptr<QueryExecutionTree> rightop, size_t inputCol) const {
-  return bindLeftOrRightSide(std::move(rightop), inputCol, false);
+template <size_t SUB_WIDTH, size_t SIDE_WIDTH>
+std::pair<TransitivePath::Map, std::vector<Id>>
+TransitivePath::setupMapAndNodes(const IdTable& sub,
+                                 const TransitivePathSide& startSide,
+                                 const TransitivePathSide& targetSide,
+                                 const IdTable& startSideTable) const {
+  std::vector<Id> nodes;
+  Map edges = setupEdgesMap<SUB_WIDTH>(sub, startSide, targetSide);
+
+  // Bound -> var|id
+  nodes = setupNodesVector<SIDE_WIDTH>(startSideTable,
+                                       startSide.treeAndCol.value().second);
+
+  return std::make_pair(edges, nodes);
 }
 
 // _____________________________________________________________________________
-std::shared_ptr<TransitivePath> TransitivePath::bindLeftOrRightSide(
-    std::shared_ptr<QueryExecutionTree> leftOrRightOp, size_t inputCol,
-    bool isLeft) const {
-  // Enforce required sorting of `leftOrRightOp`.
-  leftOrRightOp = QueryExecutionTree::createSortedTree(std::move(leftOrRightOp),
-                                                       {inputCol});
-  // Create a copy of this.
-  //
-  // NOTE: The RHS used to be `std::make_shared<TransitivePath>()`, which is
-  // wrong because it first calls the copy constructor of the base class
-  // `Operation`, which  would then ignore the changes in `variableColumnMap_`
-  // made below (see `Operation::getInternallyVisibleVariableColumns` and
-  // `Operation::getExternallyVariableColumns`).
-  std::shared_ptr<TransitivePath> p = std::make_shared<TransitivePath>(
-      getExecutionContext(), _subtree, _leftIsVar, _rightIsVar, _leftSubCol,
-      _rightSubCol, _leftValue, _rightValue, Variable{_leftColName},
-      Variable{_rightColName}, _minDist, _maxDist);
-  if (isLeft) {
-    p->_leftSideTree = leftOrRightOp;
-    p->_leftSideCol = inputCol;
+template <size_t SUB_WIDTH>
+std::pair<TransitivePath::Map, std::vector<Id>>
+TransitivePath::setupMapAndNodes(const IdTable& sub,
+                                 const TransitivePathSide& startSide,
+                                 const TransitivePathSide& targetSide) const {
+  std::vector<Id> nodes;
+  Map edges = setupEdgesMap<SUB_WIDTH>(sub, startSide, targetSide);
+
+  // id -> var|id
+  if (startSide.isBound()) {
```

Comment:
Do you mean `isConstantId` instead of `isBound()` (isBound is a more generic name).

Reply:
In this case we already know that the startSide can not be bound, because the function was called without the `startSideTable` argument. Then, if the start side is bound, it can only be an Id.

But I have contemplated adding a `isConstantId` function to the `TransitivePathSide` struct. There are a few places where this check is necessary, but it might clarify things in other instances, such as this one. What do you think?

Reply:
Nevermind, `isConstantId` is the inverse to the function `isVariable`. So calling `!isVariable()` has the same effect.